File: frontend/methods/euler.js
import {

    dS_dt,
    dI_dt,
    dR_dt

} from "../model/sir.js"


function euler(

S0,
I0,
R0,

h,
tf,

beta,
gamma

){

    //Iniciales
    let S=S0;

    let I=I0;

    let R=R0;

    let n = tf / h;


    //Vectores
    let t_values = [];

    let S_values = [];

    let I_values = [];

    let R_values = [];


    //Simulación
    for(let i=0;i<n;i++){

        //Guardar
        t_values.push(i*h);

        S_values.push(S);

        I_values.push(I);

        R_values.push(R);


        //Derivadas
        let dS = dS_dt(
            S,
            I,
            beta
        );

        let dI = dI_dt(
            S,
            I,
            beta,
            gamma
        );

        let dR = dR_dt(
            I,
            gamma
        );


        //Euler
        S = S + h*dS;

        I = I + h*dI;

        R = R + h*dR;

    }


    return{

        t:t_values,

        S:S_values,

        I:I_values,

        R:R_values

    }

}


export{

    euler

}

File: euler.py
def euler(S0, I0, R0, h, n, beta, gamma, dS_dt, dI_dt, dR_dt):
    
    #Vectores vacíos
    S = [0] * n
    I = [0] * n
    R = [0] * n

    #Condiciones iniciales
    S[0] = S0
    I[0] = I0
    R[0] = R0

    #Método de Euler
    for i in range(n - 1):

        #Pendientes actuales
        dS = dS_dt(S[i], I[i], beta)
        dI = dI_dt(S[i], I[i], beta, gamma)
        dR = dR_dt(I[i], gamma)

        #Actualizar valores
        S[i + 1] = S[i] + h * dS
        I[i + 1] = I[i] + h * dI
        R[i + 1] = R[i] + h * dR

    return S, I, R